File: python/7-distancevector.py
class Router:
    def __init__(self, id):
        self.id = id
        self.routing_table = {}

    def add_link(self, dest, cost):
        self.routing_table[dest] = cost


def initialize_routing_table(routers, num_routers, source):
    for i in range(num_routers):
        if i == source:
            routers[i] = Router(i)
            routers[i].add_link(i, 0)  # Distance to itself is 0
        else:
            routers[i] = Router(i)
            routers[i].add_link(i, float('inf'))  # Initialize to infinity


def distance_vector_routing(routers, num_routers):
    updated = True
    while updated:
        updated = False
        for i in range(num_routers):
            for dest, cost in routers[i].routing_table.items():
                for neighbor_dest, neighbor_cost in routers[dest].routing_table.items():
                    if routers[i].routing_table.get(neighbor_dest, float('inf')) > cost + neighbor_cost:
                        routers[i].routing_table[neighbor_dest] = cost + neighbor_cost
                        updated = True


def print_routing_table(routers, num_routers, source):
    print(f"Routing Table for Source Router {source}:")
    for i in range(num_routers):
        print(f"Destination: {i} | Cost: {routers[source].routing_table[i]} | Next Hop: ", end="")
        if routers[source].routing_table[i] == 0:
            print("Self")
        elif routers[source].routing_table[i] == float('inf'):
            print("Not Reachable")
        else:
            next_hop = i
            while routers[source].routing_table[next_hop] != 0:
                for link_dest, link_cost in routers[next_hop].routing_table.items():
                    if routers[source].routing_table[next_hop] - link_cost == routers[source].routing_table[i]:
                        next_hop = link_dest
                        break
            print(next_hop)
    print()


def main():
    num_routers = 5
    routers = [None] * num_routers

    source = int(input("Enter the source router ID: "))
    initialize_routing_table(routers, num_routers, source)

    # Define links between routers (source, destination, cost)
    routers[0].add_link(1, 2)
    routers[0].add_link(2, 3)
    routers[1].add_link(0, 2)
    routers[1].add_link(2, 1)
    routers[1].add_link(3, 3)
    routers[2].add_link(0, 3)
    routers[2].add_link(1, 1)
    routers[2].add_link(3, 1)
    routers[2].add_link(4, 5)
    routers[3].add_link(1, 3)
    routers[3].add_link(2, 1)
    routers[3].add_link(4, 2)
    routers[4].add_link(2, 5)
    routers[4].add_link(3, 2)

    # Perform distance vector routing
    distance_vector_routing(routers, num_routers)

    # Print routing tables for all routers
    for i in range(num_routers):
        print_routing_table(routers, num_routers, i)


if __name__ == "__main__":
    main()

File: c/7-distancevector.c
#include <stdio.h>
#include <limits.h>

#define NUM_ROUTERS 5

// Structure to represent a router
typedef struct {
    int id;
    int routing_table[NUM_ROUTERS][2]; // [destination][cost, next_hop]
} Router;

// Function to initialize routing table for all routers
void init_routing_table(Router routers[], int source) {
    for (int i = 0; i < NUM_ROUTERS; ++i) {
        routers[i].id = i;
        for (int j = 0; j < NUM_ROUTERS; ++j) {
            if (i == source) {
                routers[i].routing_table[j][0] = (i == j) ? 0 : INT_MAX; // Distance to itself is 0, others are infinity
                routers[i].routing_table[j][1] = (i == j) ? i : -1; // Next hop is itself for self, -1 for others
            } else {
                routers[i].routing_table[j][0] = INT_MAX; // Initialize to infinity
                routers[i].routing_table[j][1] = -1; // Initialize next hop to -1
            }
        }
    }
}

// Function to perform distance vector routing
void distance_vector_routing(Router routers[]) {
    int updated;
    do {
        updated = 0;
        for (int i = 0; i < NUM_ROUTERS; ++i) {
            for (int j = 0; j < NUM_ROUTERS; ++j) {
                for (int k = 0; k < NUM_ROUTERS; ++k) {
                    if (routers[i].routing_table[j][0] > routers[i].routing_table[k][0] + routers[k].routing_table[j][0]) {
                        routers[i].routing_table[j][0] = routers[i].routing_table[k][0] + routers[k].routing_table[j][0];
                        routers[i].routing_table[j][1] = k;
                        updated = 1;
                    }
                }
            }
        }
    } while (updated);
}

// Function to print routing table of all routers
void print_routing_table(Router routers[], int source) {
    printf("Routing Table for Source Router %d:\n", source);
    for (int i = 0; i < NUM_ROUTERS; ++i) {
        printf("Destination: %d | Cost: %d | Next Hop: ", i, routers[source].routing_table[i][0]);
        if (routers[source].routing_table[i][0] == 0) {
            printf("Self\n");
        } else if (routers[source].routing_table[i][0] == INT_MAX) {
            printf("Not Reachable\n");
        } else {
            int next_hop = i;
            while (routers[source].routing_table[next_hop][0] != 0) {
                next_hop = routers[source].routing_table[next_hop][1];
            }
            printf("%d\n", next_hop);
        }
    }
    printf("\n");
}

int main() {
    Router routers[NUM_ROUTERS];
    int source;

    printf("Enter the source router ID: ");
    scanf("%d", &source);

    // Initialize routing tables for all routers
    init_routing_table(routers, source);

    // Define links between routers (source, destination, cost)
    routers[0].routing_table[1][0] = 2;
    routers[0].routing_table[2][0] = 3;
    routers[1].routing_table[0][0] = 2;
    routers[1].routing_table[2][0] = 1;
    routers[1].routing_table[3][0] = 3;
    routers[2].routing_table[0][0] = 3;
    routers[2].routing_table[1][0] = 1;
    routers[2].routing_table[3][0] = 1;
    routers[2].routing_table[4][0] = 5;
    routers[3].routing_table[1][0] = 3;
    routers[3].routing_table[2][0] = 1;
    routers[3].routing_table[4][0] = 2;
    routers[4].routing_table[2][0] = 5;
    routers[4].routing_table[3][0] = 2;

    // Perform distance vector routing
    distance_vector_routing(routers);

    // Print routing tables for all routers
    for (int i = 0; i < NUM_ROUTERS; ++i) {
        print_routing_table(routers, i);
    }

    return 0;
}
